### ISSUE TITLE
Sikuli Timeout

### ISSUE THREAD

**Issue by seekM** [dictation-toolbox/Caster#464]
From time to time after executing a Sikuli command via Caster I get the errors below. Afer that Sikuli commands don't work anymore. Unfortunately, I can't give an example command which always generates this error. I don't experience a consistency in this.

Error Natlink window:
```
<class 'socket.timeout'>
timed out
<traceback object at 0x0EF108A0>
  File "C:\dev\Natlink-User\Caster\castervoice\asynch\sikuli\sikuli.py", line 52, in execute
    fn()

  File "C:\dev\Python27\Lib\xmlrpclib.py", line 1243, in __call__
    return self.__send(self.__name, args)

  File "C:\dev\Python27\Lib\xmlrpclib.py", line 1602, in __request
    verbose=self.__verbose

  File "C:\dev\Python27\Lib\xmlrpclib.py", line 1283, in request
    return self.single_request(host, handler, request_body, verbose)

  File "C:\dev\Python27\Lib\xmlrpclib.py", line 1313, in single_request
    response = h.getresponse(buffering=True)

  File "C:\dev\Python27\Lib\httplib.py", line 1121, in getresponse
    response.begin()

  File "C:\dev\Python27\Lib\httplib.py", line 438, in begin
    version, status, reason = self._read_status()

  File "C:\dev\Python27\Lib\httplib.py", line 394, in _read_status
    line = self.fp.readline(_MAXLINE + 1)

  File "C:\dev\Python27\Lib\socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
```

Error Sikuli server window:
```
Exception happened during processing of request from ('127.0.0.1', 60764)
Traceback (most recent call last):
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\SocketServer.py", line 295, in _handle_request_noblock
    self.process_request(request, client_address)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\SocketServer.py", line 659, in __init__
    self.finish()
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\SocketServer.py", line 718, in finish
    self.wfile.close()
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 2009, in close
    self.flush()
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 2036, in flush
    self._sock.sendall(chunk)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 1484, in send
    return super(ChildSocket, self).send(data)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 382, in handle_exception
    return method_or_function(*args, **kwargs)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 382, in handle_exception
    return method_or_function(*args, **kwargs)
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 1200, in send
    self._handle_channel_future(future, "send")
  File "C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib\_socket.py", line 384, in handle_exception
    raise _map_exception(jlx)
error: [Errno 10054] Socket closed
```
Anyone who experiences this as well? Any idea?

Specs:
Sikuli 1.4
java version "1.8.0_211"
Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.211-b12, mixed mode)

**Comment by LexiconCode** [comment]
Yes experiences 100% of the time on one of my machines. It's good to know that it can be Intermittent.

**Comment by LexiconCode** [comment]
However the error differs in the Sikuli server window:
```
Exception happened during processing of request from ('127.0.0.1', 45432)
Traceback (most recent call last):
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\SocketServer.py", line 295, in _handle_request_noblock
    self.process_request(request, client_address)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\SocketServer.py", line 659, in __init__
    self.finish()
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\SocketServer.py", line 718, in finish
    self.wfile.close()
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 2009, in close
    self.flush()
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 2036, in flush
    self._sock.sendall(chunk)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 1484, in send
    return super(ChildSocket, self).send(data)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 382, in handle_exception
    return method_or_function(*args, **kwargs)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 382, in handle_exception
    return method_or_function(*args, **kwargs)
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 1200, in send
    self._handle_channel_future(future, "send")
  File "C:\Users\user\AppData\Roaming\Sikulix\Lib\_socket.py", line 384, in handle_exception
    raise _map_exception(jlx)
error: [Errno 10054] Socket closed
```

**Comment by LexiconCode** [comment]
However after deleting `C:\Users\user\AppData\Roaming\Sikulix` and replacing `jython-standalone-2.7.1.jar` 
I now experience the same error.

**Comment by LexiconCode** [comment]
Go to C:\Users\User\.caster\sikuli
open command prompt and execute the following taking into account your SikuliX install path C:\SikuliX\SikuliX_1.1.4\sikulix.jar and SikuliX script name click.sikuli

java -jar C:\SikuliX\SikuliX_1.1.4\sikulix.jar -v -r click.sikuli

Can you post output and does the Script run?

**Comment by seekM** [comment]
For my custom test.sikuli file:

```
C:\Users\User\.caster\sikuli>java -jar C:\dev\Sikuli\sikulix.jar -v -r test.sikuli
[DEBUG] IDE: starting
[DEBUG] IDE: no extensions.txt nor valid content
[DEBUG] IDE: adding extension: C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar
[DEBUG] IDE: Running: C:\dev\Sikuli\sikulix.jar
[DEBUG] IDE: AppData: C:\Users\User\AppData\Roaming\Sikulix
[DEBUG] IDE: Classpath: C:\dev\Sikuli\sikulix.jar;C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar
[DEBUG] IDE: no extensions.txt nor valid content
[39 debug] RunTimeINIT: temp folder ok: C:\Users\User\AppData\Local\Temp\Sikulix_1727709388
***** show environment for 1.1.4-SNAPSHOT-#285-2019-04-17_08:42 API
user.home: C:\Users\User
user.dir (work dir): C:\Users\User\.caster\sikuli
user.name: User
java.io.tmpdir: C:\Users\User\AppData\Local\Temp
running 64Bit(amd64) on w (10.0) from a jar
java 8 version 1.8 vm 25.211-b12 class 52.0 arch 64
app data folder: C:\Users\User\AppData\Roaming\Sikulix
executing jar: C:\dev\Sikuli\sikulix.jar
*** classpath dump sikulix
  0: file:/C:/dev/Sikuli/sikulix.jar
  1: file:/C:/Users/User/AppData/Roaming/Sikulix/Extensions/jython-standalone-2.7.1.jar
*** classpath dump end
***** show environment end
[920 debug] Runner: added: Robot [robot] text/robot
[925 debug] Runner: added: NetworkRunner [] NET
[930 debug] Runner: added: Jython [py] text/jython
[930 debug] Runner: added: Sikulix [sikuli] directory/sikulix
[930 debug] Runner: added: PowerShell [ps1] text/powershell
[1150 debug] Runner: added: JavaScript [js] text/javascript
[1150 debug] Runner: added: Text [txt] text/text
[1150 debug] ImagePath: new BundlePath:
/C:/Users/User/.caster/sikuli/test.sikuli/
[1150 debug] Jython: starting initialization
[3270 debug] RunTimeAPI: resourceLocation: (class org.sikuli.script.support.RunTime) /Lib/sikuli
[3270 debug] RunTimeAPI: resourceList: having jar: jar:file:/C:/dev/Sikuli/sikulix.jar!/Lib/sikuli
[4820 debug] RunTimeAPI: files exported: 7 from: Lib/sikuli to:
 C:\Users\User\AppData\Roaming\Sikulix\Lib\sikuli
[4820 debug] Jython: added as Jython::sys.path[0]:
C:\Users\User\AppData\Roaming\Sikulix\Lib\site-packages
[4830 debug] Jython: ***** sys.path
 0: /C:/Users/User/.caster/sikuli/test.sikuli/
 1: C:\Users\User\AppData\Roaming\Sikulix\Lib\site-packages
 2: C:\Users\User\AppData\Roaming\Sikulix\Lib
 3: C:\Users\User\AppData\Roaming\Sikulix\Extensions\Lib
 4: C:\Users\User\AppData\Roaming\Sikulix\Extensions\jython-standalone-2.7.1.jar\Lib
[4840 debug] Jython: ready: version 2.7.1
[debug] Jython: SikulixForJython: init: starting
[debug] Jython: added as Jython::sys.path[0]:
C:\Users\User\AppData\Roaming\Sikulix\Lib\site-packages
[debug] Jython: Default ImagePath: null
[debug] Jython: SikulixForJython: init: success
[debug] Jython: sikuli: Sikuli: starting init
[debug] Screen: initScreens: starting
[debug] Screen: Accessing: GraphicsEnvironment.getLocalGraphicsEnvironment()
[debug] Screen: Accessing: GraphicsEnvironment.getLocalGraphicsEnvironment().getScreenDevices()
[debug] Screen: ScreenDevice 0 has (0,0) --- will be primary Screen(0)
[debug] Screen: Monitor 0: (0, 0) 1920 x 1080
[debug] Screen: initScreens: ending
[debug] Jython: use as default region: R[0,0 1920x1080]@S(0)
[debug] Jython: sikuli: Sikuli: ending init
[debug] Jython: use as default region: R[0,0 1920x1080]@S(0)
[debug] RunTimeAPI: ***** final cleanup at System.exit() *****
[debug] FileManager: deleteFileOrFolder:
C:\Users\User\AppData\Local\Temp\Sikulix_1727709388
IDE terminated: returned: 0
```

**Comment by LexiconCode** [comment]
Does the script run clicking on the target?

**Comment by LexiconCode** [comment]
I believe-figured out the conditions to make this repeatable. Therefore I'll need your help to verify.
 The crash occurs when Sikulix cannot find a match on the screen. I created a Sikulix command to click on a desktop shortcut. Zero issues with timing out when Sikulix successful finding and clicking on the icon. Cover up the icon was another window the timeout occurs every time. @seekM can you verify?

**Comment by LexiconCode** [comment]
I tested running the script with failing conditions in Sikulix IDE. It produces the following error.
The error seems very ambiguous and I'm not sure why it's creating a Traceback for something that is expected behavior.
```
[error] script [ start ] stopped with error at line --unknown--
[error] Error caused by: Traceback (most recent call last): File "C:\Users\user\.caster\sikuli\start.sikuli\start.py", line 4, in <module> export_test() File "C:\Users\user\.caster\sikuli\start.sikuli\start.py", line 3, in export_test click(Pattern("1555705743103.png").targetOffset(-1,-15)) Line 2341, in file Region.java 
at org.sikuli.script.Region.wait(Region.java:2341)
at org.sikuli.script.Region.wait(Region.java:2359)
at org.sikuli.script.Region.getLocationFromTarget(Region.java:3386)
at org.sikuli.script.Region.click(Region.java:4046)
at org.sikuli.script.Region.click(Region.java:4022)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
at java.lang.reflect.Method.invoke(Unknown Source)
org.sikuli.script.FindFailed: FindFailed: 1555705743103.png: (147x176) seen at (126, 147) with 1.00 in R[0,0 3840x2160]@S(0) Line 2341, in file Region.java 
```

**Comment by seekM** [comment]
I can confirm that Sikuli not finding a match on the screen always (whenever I tried this scenario) generates the error in the Natlink window. However, the error in the Sikuli server window is not generated.

I'm pretty sure though that error also occured for me when Sikuli should have found a match on the screen. I was testing a script which searches for the "Profile" button in DNS' bar on the top right which is always visible. I executed the script a few times in a row and it worked, then suddenly it stopped working and produced the error messages. Maybe that was a bug in Sikuli though which thought there was no match? I noticed that Sikuli's match function isn't perfect, for example a picture of "commented" on this page matches "comment" which is displayed in the empty write text field on this page.

Maybe Caster can deal with this behaviour of Sikuli by catching this error and doing what is necessary (e.g. restart Sikuli?)?

**Comment by seekM** [comment]
Two things I noticed: When calling a Sikuli script by voice which will fail, e.g. because the image searched for is not visible on the screen, two things can happen
- An error message is shown in both the Natlink and Sikuli window. If this is the case, subsequent Sikuli voice commands continue to work.
- An error message is shown in the Natlink window only. Subsequent Sikuli voice commands don't work anymore. However, if you press a key and the Sikuli windows is focused the error message appears in the Siuli window. After that subsequent Sikuli voice commands work again. Idea: In this case, simulate a keystroke. I tried [this](https://stackoverflow.com/a/11910555/1219728), but `ctypes` doesn't work with Jython.
